compiler/rustc_lint/src/duplicate_trait.rs

```diff
@@ -0,0 +1,65 @@
+use crate::hir;
+
+use crate::{lints::DuplicateTraitDiag, LateContext, LateLintPass};
+
+declare_lint! {
+    /// The `lint_duplicate_trait` lints repetition of traits.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// fn foo(_: &(dyn MyTrait + Send + Send>) {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Duplicate trait `Send` in trait object.
+    pub DUPLICATE_TRAIT,
+    Warn,
+    "duplicate trait constraint in trait object"
+}
+
+declare_lint_pass!(DuplicateTrait => [DUPLICATE_TRAIT]);
+
+impl<'tcx> LateLintPass<'tcx> for DuplicateTrait {
+    fn check_ty(&mut self, cx: &LateContext<'tcx>, ty: &'tcx hir::Ty<'tcx>) {
+        let hir::TyKind::Ref(
+            ..,
+            hir::MutTy {
+                ty: hir::Ty {
+                    kind: hir::TyKind::TraitObject(bounds, ..),
+                    ..
+                },
+                ..
+            }
+        ) = ty.kind else { return; };
+
+        let mut bounds = (*bounds).to_owned();
+
+        bounds.sort_unstable_by_key(|b| b.trait_ref.trait_def_id());
```

Comment:
DefIds are not stable across compilations and should not be used for sorting like here. Instead of what you're doing right now, you should use a `FxHashSet` and iterate through the bounds (in source order) and insert their def ids into the set. If a def id is already in the set, emit the lint.

Reply:
Will make that change. Out of curiosity, why does the fact they're unstable across compilations matter? The only aim of the sort is to group like `DefId`s together, which should still be valid given it is within one compilation right?